if miss type `nil nil` instead of `nil, nil` then wrong column number reported by parser

```
aqua --input aqua --output generated/air --import node_modules --air
Aqua JS: node /home/dz/.nvm/versions/node/v16.5.0/lib/node_modules/@fluencelabs/aqua/aqua.js -m node_modules --input aqua --output generated/air --import node_modules --air
Aqua JS: 2021.09.08 16:20:22 [ERROR] Example .aqua:73:13
70         example_dht_record <- Example .get_random_value_unwrap(example_replica_set_id)
71         on example_dht_record.peer_id:
72             Example example_dht_record.service_id!
73             result <- Example .get_messages(nil,nil, nil nil, last_message_seen)
               ^======================================================================
               Syntax error, expected: OneOfStr(3463,List(<-, catch, co, else, for, if, on, otherwise, par, try)), InRange(3463,A,Z), OneOfStr(3470,List(<<-)), InRange(3470,:,:), InRange(3470,=,=), InRange(3511,),))
74  --           panic_number_nil(result.ret_code, 200, resultss.err_msg)
75     <- result.body!0
76
```

--- a/aqua-src/import.aqua
+++ b/aqua-src/import.aqua
@@ -3,4 +3,4 @@ service GetStr("multiret-test"):
 
 
 func foo_wrapper():
-    GetStr.retStr(nil nil)
+    GetStr.retStr(nil, nil)

--- a/aqua-src/import.aqua
+++ b/aqua-src/import.aqua
@@ -1,9 +1,6 @@
-module Import
+service GetStr("multiret-test"):
+    retStr(s: *string, a: *string)
 
-use "export"
 
-export foo_wrapper
-
-func foo_wrapper() -> string:
-    z <- Export.foo()
-    <- z
+func foo_wrapper():
+    GetStr.retStr(nil nil)

--- a/aqua-src/import.aqua
+++ b/aqua-src/import.aqua
@@ -4,3 +4,8 @@ service GetStr("multiret-test"):
 
 func foo_wrapper():
     GetStr.retStr(nil, nil)
+
+func invalid_append():
+    stream: *string
+    variable <- Op.identity("hello")
+    stream <- variable

--- a/aqua-src/ret.aqua
+++ b/aqua-src/ret.aqua
@@ -1,6 +1,6 @@
 module Ret
 
-import Service from "service.aqua"
+import Service from "service"
 
 use "error.aqua"
 export GetStr, multiReturnFunc, Service as S

--- a/cli/.jvm/src/main/scala/aqua/Test.scala
+++ b/cli/.jvm/src/main/scala/aqua/Test.scala
@@ -18,10 +18,9 @@ object Test extends IOApp.Simple {
         .compileFilesTo[IO](
           Path("./aqua-src/"),
           List(Path("./aqua")),
-          Path("./target"),
+          Option(Path("./target")),
           TypeScriptBackend,
-          TransformConfig(),
-          false
+          TransformConfig()
         )
         .map {
           case Validated.Invalid(errs) =>

--- a/cli/.jvm/src/test/scala/WriteFileSpec.scala
+++ b/cli/.jvm/src/test/scala/WriteFileSpec.scala
@@ -20,7 +20,7 @@ class WriteFileSpec extends AnyFlatSpec with Matchers {
 
     val bc = TransformConfig()
     AquaPathCompiler
-      .compileFilesTo[IO](src, List.empty, targetTs, TypeScriptBackend, bc, false)
+      .compileFilesTo[IO](src, List.empty, Option(targetTs), TypeScriptBackend, bc)
       .unsafeRunSync()
       .leftMap { err =>
         println(err)
@@ -32,7 +32,7 @@ class WriteFileSpec extends AnyFlatSpec with Matchers {
     Files[IO].deleteIfExists(targetTsFile).unsafeRunSync()
 
     AquaPathCompiler
-      .compileFilesTo[IO](src, List.empty, targetJs, JavaScriptBackend, bc, false)
+      .compileFilesTo[IO](src, List.empty, Option(targetJs), JavaScriptBackend, bc)
       .unsafeRunSync()
       .leftMap { err =>
         println(err)
@@ -44,7 +44,7 @@ class WriteFileSpec extends AnyFlatSpec with Matchers {
     Files[IO].deleteIfExists(targetJsFile).unsafeRunSync()
 
     AquaPathCompiler
-      .compileFilesTo[IO](src, List.empty, targetAir, AirBackend, bc, false)
+      .compileFilesTo[IO](src, List.empty, Option(targetAir), AirBackend, bc)
       .unsafeRunSync()
       .leftMap { err =>
         println(err)

--- a/cli/src/main/scala/aqua/AppOps.scala
+++ b/cli/src/main/scala/aqua/AppOps.scala
@@ -44,7 +44,7 @@ object AppOps {
       )
   }
 
-  def checkOutput[F[_]: Monad: Files](pathStr: String): F[ValidatedNec[String, Path]] = {
+  def checkOutput[F[_]: Monad: Files](pathStr: String): F[ValidatedNec[String, Option[Path]]] = {
     val p = Path(pathStr)
     Files[F]
       .exists(p)
@@ -54,10 +54,10 @@ object AppOps {
             if (isFile) {
               Validated.invalidNec(s"Output path should be a directory. Current: '$p'")
             } else
-              Validated.validNec(p)
+              Validated.validNec(Option(p))
           }
         else
-          Files[F].createDirectories(p).map(_ => Validated.validNec(p))
+          Files[F].createDirectories(p).map(_ => Validated.validNec(Option(p)))
       }
   }
 
@@ -88,8 +88,11 @@ object AppOps {
       )
       .map(s => checkPath[F](s))
 
-  def outputOpts[F[_]: Monad: Files]: Opts[F[ValidatedNec[String, Path]]] =
-    Opts.option[String]("output", "Path to the output directory. Will be created if not exists", "o").map(s => checkOutput[F](s))
+  def outputOpts[F[_]: Monad: Files]: Opts[F[ValidatedNec[String, Option[Path]]]] =
+    Opts.option[String]("output", "Path to the output directory. Will be created if not exists", "o")
+      .map(s => Option(s))
+      .withDefault(None)
+      .map(_.map(checkOutput[F]).getOrElse(Validated.validNec[String, Option[Path]](None).pure[F]))
 
   def importOpts[F[_]: Monad: Files]: Opts[F[ValidatedNec[String, List[Path]]]] =
     Opts
@@ -167,7 +170,7 @@ object AppOps {
 
   val dryOpt: Opts[Boolean] =
     Opts
-      .flag("dry", "Generate .air file instead of typescript")
+      .flag("dry", "Checks if compilation is succeeded, without output")
       .map(_ => true)
       .withDefault(false)
 

--- a/cli/src/main/scala/aqua/AquaCli.scala
+++ b/cli/src/main/scala/aqua/AquaCli.scala
@@ -86,32 +86,43 @@ object AquaCli extends IOApp with Logging {
             i.andThen { input =>
               o.andThen { output =>
                 imp.map { imports =>
-                  AquaPathCompiler
-                    .compileFilesTo[F](
-                      input,
-                      imports,
-                      output,
-                      targetToBackend(target),
-                      bc,
-                      isDryRun
-                    )
+                  if (output.isEmpty && !isDryRun)
+                    Validated.invalidNec(
+                      "Output path should be specified ('--output' or '-o'). " +
+                        "Add '--dry' to check compilation without output")
+                      .pure[F]
+                  else {
+                    val resultOutput = if (isDryRun) {
+                      None
+                    } else {
+                      output
+                    }
+                    AquaPathCompiler
+                      .compileFilesTo[F](
+                        input,
+                        imports,
+                        resultOutput,
+                        targetToBackend(target),
+                        bc
+                      )
                   }
                 }
               }
-            }.flatMap {
-              case Validated.Invalid(errs) =>
-                errs.map(logger.error(_))
-                ExitCode.Error.pure[F]
-              case Validated.Valid(result) =>
-                result.map {
-                  case Validated.Invalid(errs) =>
-                    errs.map(logger.error(_))
-                    ExitCode.Error
-                  case Validated.Valid(results) =>
-                    results.map(logger.info(_))
-                    ExitCode.Success
-                }
             }
+          }.flatMap {
+            case Validated.Invalid(errs) =>
+              errs.map(logger.error(_))
+              ExitCode.Error.pure[F]
+            case Validated.Valid(result) =>
+              result.map {
+                case Validated.Invalid(errs) =>
+                  errs.map(logger.error(_))
+                  ExitCode.Error
+                case Validated.Valid(results) =>
+                  results.map(logger.info(_))
+                  ExitCode.Success
+              }
+          }
         }
     }
   }

--- a/cli/src/main/scala/aqua/AquaPathCompiler.scala
+++ b/cli/src/main/scala/aqua/AquaPathCompiler.scala
@@ -41,7 +41,7 @@ object AquaPathCompiler extends Logging {
             val nat = new (Span.F ~> FileSpan.F) {
               override def apply[A](span: Span.F[A]): FileSpan.F[A] = {
                 (
-                  FileSpan(id.file.fileName.toString, Eval.later(LocationMap(source)), span._1),
+                  FileSpan(id.file.absolute.toString, Eval.later(LocationMap(source)), span._1),
                   span._2
                 )
               }

--- a/cli/src/main/scala/aqua/AquaPathCompiler.scala
+++ b/cli/src/main/scala/aqua/AquaPathCompiler.scala
@@ -11,7 +11,7 @@ import cats.data.*
 import cats.syntax.functor.*
 import cats.syntax.applicative.*
 import cats.syntax.show.*
-import cats.{Eval, Monad, Show, ~>, Applicative}
+import cats.{~>, Applicative, Eval, Monad, Show}
 import scribe.Logging
 import fs2.io.file.{Files, Path}
 import aqua.parser.lift.{LiftParser, Span}
@@ -23,13 +23,20 @@ import aqua.parser.Parser
 
 object AquaPathCompiler extends Logging {
 
+  /**
+   * @param srcPath path to aqua sources
+   * @param imports additional paths to possible aqua imports
+   * @param targetPath path where compiled files will be created. Creates no output if empty
+   * @param backend creates output (TS, JS, ...) from a model
+   * @param transformConfig transformation configuration for a model
+   * @return errors or result messages
+   */
   def compileFilesTo[F[_]: AquaIO: Monad: Files](
     srcPath: Path,
     imports: List[Path],
-    targetPath: Path,
+    targetPath: Option[Path],
     backend: Backend,
-    bodyConfig: TransformConfig,
-    isDryRun: Boolean
+    transformConfig: TransformConfig
   ): F[ValidatedNec[String, Chain[String]]] = {
     import ErrorRendering.showError
     val sources = new AquaFileSources[F](srcPath, imports)
@@ -51,13 +58,15 @@ object AquaPathCompiler extends Logging {
           }
         },
         backend,
-        bodyConfig,
-        {if (isDryRun) dry[F] else sources.write(targetPath)}
+        transformConfig,
+        targetPath.map(sources.write).getOrElse(dry[F])
       )
       .map(_.leftMap(_.map(_.show)))
   }
 
-  def dry[F[_]: Applicative](ac: AquaCompiled[FileModuleId]): F[Seq[Validated[AquaFileError, String]]] =
+  def dry[F[_]: Applicative](
+    ac: AquaCompiled[FileModuleId]
+  ): F[Seq[Validated[AquaFileError, String]]] =
     Seq(
       Validated.valid[AquaFileError, String](
         s"Source ${ac.sourceId.file}: compilation OK"

--- a/cli/src/main/scala/aqua/ErrorRendering.scala
+++ b/cli/src/main/scala/aqua/ErrorRendering.scala
@@ -24,7 +24,7 @@ object ErrorRendering {
   def expectationToString(expectation: Expectation): String = {
     // TODO: match all expectations
     expectation match {
-      case wc@WithContext(str, offset) => str
+      case wc@WithContext(str, exp) => s"$str (${expectationToString(exp)})"
       case InRange(offset, lower, upper) =>
         if (lower == upper)
           s"Expected symbol '${betterSymbol(lower)}'"

--- a/cli/src/main/scala/aqua/ErrorRendering.scala
+++ b/cli/src/main/scala/aqua/ErrorRendering.scala
@@ -13,35 +13,35 @@ import cats.parse.Parser.Expectation.*
 
 object ErrorRendering {
 
-  def expectationToString(expectation: Expectation, locationMap: Eval[LocationMap], currentOffset: Int): String = {
-    // add column number if it is not on the current offset
-    def makeMsg(msg: String, offset: Int): String = {
-      if (offset == currentOffset) msg
-      else {
-        val focus = Span(offset, offset + 1).focus(locationMap, 0)
-        focus.map(f => s"$msg on ${f.line._1}:${f.column}").getOrElse(msg)
-      }
+  def betterSymbol(symbol: Char): String = {
+    symbol match {
+      case ' ' => "whitespace"
+      case '\t' => "tabulation"
+      case c => c.toString
     }
+  }
 
+  def expectationToString(expectation: Expectation): String = {
     // TODO: match all expectations
     expectation match {
-      case wc@WithContext(str, offset) => makeMsg(str, wc.offset)
+      case wc@WithContext(str, offset) => str
       case InRange(offset, lower, upper) =>
         if (lower == upper)
-          makeMsg(s"Expected symbol '$lower'", offset)
+          s"Expected symbol '${betterSymbol(lower)}'"
         else
-          makeMsg(s"Expected symbols from '$lower' to '$upper'", offset)
+          s"Expected symbols from '${betterSymbol(lower)}' to '${betterSymbol(upper)}'"
       case OneOfStr(offset, strs) =>
-        makeMsg(s"Expected one of these strings: ${strs.map(s => s"'$s'").mkString(", ")}", offset)
+        s"Expected one of these strings: ${strs.map(s => s"'$s'").mkString(", ")}"
       case e => "Expected: " + e.toString
     }
   }
 
-  def showForConsole(span: FileSpan, messages: List[String]): String =
+  def showForConsole(errorType: String, span: FileSpan, messages: List[String]): String =
     span
       .focus(3)
       .map(
         _.toConsoleStr(
+          errorType,
           messages,
           Console.RED
         )
@@ -53,46 +53,50 @@ object ErrorRendering {
   implicit val showError: Show[AquaError[FileModuleId, AquaFileError, FileSpan.F]] = Show.show {
     case ParserErr(err) =>
       err match {
-        case BlockIndentError(indent, message) => showForConsole(indent._1, message :: Nil)
-        case FuncReturnError(point, message) => showForConsole(point._1, message :: Nil)
+        case BlockIndentError(indent, message) => showForConsole("Syntext error", indent._1, message :: Nil)
+        case FuncReturnError(point, message) => showForConsole("Syntext error", point._1, message :: Nil)
         case LexerError((span, e)) =>
-          span
-            .focus(3)
-            .map { spanFocus =>
-              val errorMessages = e.expected.map(exp => expectationToString(exp, span.locationMap, span.span.startIndex))
-              spanFocus.toConsoleStr(
-                s"Syntax error: ${errorMessages.head}" :: errorMessages.tail.map(t => "OR " + t),
-                Console.RED
-              )
-            }
-            .getOrElse(
-              "(offset is beyond the script, syntax errors) " + Console.RED + e.expected.toList
-                .mkString(", ")
-            ) + Console.RESET + "\n"
+          e.expected.toList.groupBy(_.offset).map { case (offset, exps) =>
+            val localSpan = Span(offset, offset + 1)
+            val msg = FileSpan(span.name, span.locationMap, localSpan).focus(0)
+              .map { spanFocus =>
+                val errorMessages = exps.map(exp => expectationToString(exp))
+                spanFocus.toConsoleStr(
+                  "Syntext error",
+                  s"${errorMessages.head}" :: errorMessages.tail.map(t => "OR " + t),
+                  Console.RED
+                )
+              }
+              .getOrElse(
+                "(offset is beyond the script, syntax errors) " + Console.RED + e.expected.toList
+                  .mkString(", ")
+              ) + Console.RESET
+            (offset, msg)
+          }.toList.sortBy(_._1).map(_._2).reverse.mkString("\n")
       }
     case SourcesErr(err) =>
       Console.RED + err.showForConsole + Console.RESET
     case ResolveImportsErr(_, token, err) =>
       val span = token.unit._1
-      showForConsole(span, s"Cannot resolve imports: ${err.showForConsole}" :: Nil)
+      showForConsole("Cannot resolve imports", span, s"${err.showForConsole}" :: Nil)
 
     case ImportErr(token) =>
       val span = token.unit._1
-      showForConsole(span, s"Cannot resolve import" :: Nil)
+      showForConsole("Cannot resolve import", span, "Cannot resolve import" :: Nil)
     case CycleError(modules) =>
       s"Cycle loops detected in imports: ${modules.map(_.file.fileName)}"
     case CompileError(err) =>
       err match {
         case RulesViolated(token, message) =>
           token.unit._1
-            .focus(2)
-            .map(_.toConsoleStr(message :: Nil, Console.CYAN))
-            .getOrElse("(Dup error, but offset is beyond the script)") + "\n"
+            .focus(0)
+            .map(_.toConsoleStr("Error", message :: Nil, Console.CYAN))
+            .getOrElse("(Dup error, but offset is beyond the script)")
         case HeaderError(token, message) =>
           token.unit._1
-            .focus(2)
-            .map(_.toConsoleStr(message :: Nil, Console.CYAN))
-            .getOrElse("(Dup error, but offset is beyond the script)") + "\n"
+            .focus(0)
+            .map(_.toConsoleStr("Header error", message :: Nil, Console.CYAN))
+            .getOrElse("(Dup error, but offset is beyond the script)")
         case WrongAST(ast) =>
           s"Semantic error"
 

--- a/cli/src/main/scala/aqua/ErrorRendering.scala
+++ b/cli/src/main/scala/aqua/ErrorRendering.scala
@@ -53,8 +53,8 @@ object ErrorRendering {
   implicit val showError: Show[AquaError[FileModuleId, AquaFileError, FileSpan.F]] = Show.show {
     case ParserErr(err) =>
       err match {
-        case BlockIndentError(indent, message) => showForConsole("Syntext error", indent._1, message :: Nil)
-        case FuncReturnError(point, message) => showForConsole("Syntext error", point._1, message :: Nil)
+        case BlockIndentError(indent, message) => showForConsole("Syntax error", indent._1, message :: Nil)
+        case FuncReturnError(point, message) => showForConsole("Syntax error", point._1, message :: Nil)
         case LexerError((span, e)) =>
           e.expected.toList.groupBy(_.offset).map { case (offset, exps) =>
             val localSpan = Span(offset, offset + 1)
@@ -62,7 +62,7 @@ object ErrorRendering {
               .map { spanFocus =>
                 val errorMessages = exps.map(exp => expectationToString(exp))
                 spanFocus.toConsoleStr(
-                  "Syntext error",
+                  "Syntax error",
                   s"${errorMessages.head}" :: errorMessages.tail.map(t => "OR " + t),
                   Console.RED
                 )

--- a/cli/src/main/scala/aqua/files/AquaFileSources.scala
+++ b/cli/src/main/scala/aqua/files/AquaFileSources.scala
@@ -57,11 +57,8 @@ class AquaFileSources[F[_]: AquaIO: Monad: Files: Functor](
     validatedPath match {
       case Validated.Valid(importP) =>
         // if there is no `.aqua` extension, than add it
-        val renamedImport =
-          if (importP.extName != ".aqua") importP.resolveSibling(importP.fileName.toString + ".aqua")
-          else importP
         filesIO
-          .resolve(renamedImport, importFrom.prependedAll(from.file.parent))
+          .resolve(importP, importFrom.prependedAll(from.file.parent))
           .bimap(NonEmptyChain.one, FileModuleId(_))
           .value
           .map(Validated.fromEither)

--- a/parser/src/main/scala/aqua/parser/expr/CallArrowExpr.scala
+++ b/parser/src/main/scala/aqua/parser/expr/CallArrowExpr.scala
@@ -1,11 +1,12 @@
 package aqua.parser.expr
 
 import aqua.parser.Expr
-import aqua.parser.lexer.Token._
+import aqua.parser.lexer.Token.*
 import aqua.parser.lexer.{Ability, Name, Value}
 import aqua.parser.lift.LiftParser
 import cats.Comonad
-import cats.parse.{Parser => P}
+import cats.data.NonEmptyList
+import cats.parse.{Parser as P, Parser0 as P0}
 import cats.~>
 
 case class CallArrowExpr[F[_]](
@@ -26,13 +27,19 @@ case class CallArrowExpr[F[_]](
 
 object CallArrowExpr extends Expr.Leaf {
 
-  override def p[F[_]: LiftParser: Comonad]: P[CallArrowExpr[F]] =
-    ((comma(Name.p[F]) <* ` <- `).backtrack.?.with1 ~
-      ((Ability.dotted[F] <* `.`).?.with1 ~
-        Name.p[F] ~
-        comma0(Value.`value`[F].surroundedBy(`/s*`)).between(`(` <* `/s*`, `/s*` *> `)`))).map {
-      case (variables, ((ability, funcName), args)) =>
+  override def p[F[_]: LiftParser: Comonad]: P[CallArrowExpr[F]] = {
+    val variables: P0[Option[NonEmptyList[Name[F]]]] = (comma(Name.p[F]) <* ` <- `).backtrack.?
+    val ability: P0[Option[Ability[F]]] = (Ability.dotted[F] <* `.`).?
+    val functionCallWithArgs = Name.p[F]
+      ~ comma0(Value.`value`[F].surroundedBy(`/s*`)).between(`(` <* `/s*`, `/s*` *> `)`)
+
+    (variables.with1 ~
+      (ability.with1 ~ functionCallWithArgs)
+        .withContext("Only results of a function call can be written to a stream")
+      ).map {
+      case (variables, (ability, (funcName, args))) =>
         CallArrowExpr(variables.toList.flatMap(_.toList), ability, funcName, args)
     }
+  }
 
 }

--- a/parser/src/main/scala/aqua/parser/head/FilenameExpr.scala
+++ b/parser/src/main/scala/aqua/parser/head/FilenameExpr.scala
@@ -9,8 +9,13 @@ trait FilenameExpr[F[_]] extends HeaderExpr[F] {
 
   override def token: Token[F] = filename
 
-  def fileValue: String = filename.value.drop(1).dropRight(1)
+  def fileValue: String = {
+    val raw = filename.value.drop(1).dropRight(1)
+    if (raw.endsWith(".aqua"))
+      raw
+    else
+      raw + ".aqua"
+  }
 
   override def mapK[K[_]: Comonad](fk: F ~> K): FilenameExpr[K]
-    
 }

--- a/parser/src/main/scala/aqua/parser/lexer/Value.scala
+++ b/parser/src/main/scala/aqua/parser/lexer/Value.scala
@@ -1,6 +1,7 @@
 package aqua.parser.lexer
 
 import aqua.parser.Expr
+import aqua.parser.head.FilenameExpr
 import aqua.parser.lexer.Token.*
 import aqua.parser.lift.LiftParser
 import aqua.parser.lift.LiftParser.*

--- a/parser/src/main/scala/aqua/parser/lift/FileSpan.scala
+++ b/parser/src/main/scala/aqua/parser/lift/FileSpan.scala
@@ -10,15 +10,15 @@ import scala.language.implicitConversions
 case class FileSpan(name: String, locationMap: Eval[LocationMap], span: Span) {
 
   def focus(ctx: Int): Option[FileSpan.Focus] =
-    span.focus(locationMap, ctx).map(FileSpan.Focus(name, locationMap, ctx, _))
+    span.focus(locationMap, ctx, ctx).map(FileSpan.Focus(name, locationMap, ctx, _))
 }
 
 object FileSpan {
 
   case class Focus(name: String, locationMap: Eval[LocationMap], ctx: Int, spanFocus: Span.Focus) {
 
-    def toConsoleStr(msgs: List[String], onLeft: String, onRight: String = Console.RESET): String =
-      s"$name:${spanFocus.line._1 + 1}:${spanFocus.column + 1}\n" +
+    def toConsoleStr(errorType: String, msgs: List[String], onLeft: String, onRight: String = Console.RESET): String =
+      onLeft + "---- " + errorType + ": " + s"$name:${spanFocus.line._1 + 1}:${spanFocus.column + 1}" + onRight +
         spanFocus.toConsoleStr(
           msgs,
           onLeft,

--- a/parser/src/main/scala/aqua/parser/lift/FileSpan.scala
+++ b/parser/src/main/scala/aqua/parser/lift/FileSpan.scala
@@ -10,7 +10,7 @@ import scala.language.implicitConversions
 case class FileSpan(name: String, locationMap: Eval[LocationMap], span: Span) {
 
   def focus(ctx: Int): Option[FileSpan.Focus] =
-    span.focus(locationMap, ctx, ctx).map(FileSpan.Focus(name, locationMap, ctx, _))
+    span.focus(locationMap, ctx).map(FileSpan.Focus(name, locationMap, ctx, _))
 }
 
 object FileSpan {

--- a/parser/src/main/scala/aqua/parser/lift/Span.scala
+++ b/parser/src/main/scala/aqua/parser/lift/Span.scala
@@ -109,13 +109,7 @@ object Span {
       }
 
     override def wrapErr(e: P.Error): (Span, P.Error) = {
-      // Find all WithContext expectations, get last by offset.
-      // This will be the final error handled by hand.
-      val withContext = e.expected.collect {
-        case e@P.Expectation.WithContext(_, _) => e
-      }.sortBy(_.offset).headOption
-      val exp = withContext.map(e => P.Error(e.offset, NonEmptyList.one(e))).getOrElse(e)
-      (Span(exp.failedAtOffset, exp.failedAtOffset + 1), exp)
+      (Span(e.failedAtOffset, e.failedAtOffset + 1), e)
     }
   }
 

--- a/parser/src/main/scala/aqua/parser/lift/Span.scala
+++ b/parser/src/main/scala/aqua/parser/lift/Span.scala
@@ -62,6 +62,7 @@ object Span {
       val line3Length = line._3.length
       val line3Mult = if (line3Length == 0) 1 else line3Length
       val message = msgs.map(m => (" " * (line._2.length + lastNSize + 1)) + m).mkString("\n")
+
       pre.map(formatLine(_, onLeft, onRight)).mkString("\n") +
         "\n" +
         formatLN(line._1, onLeft, onRight) +

